runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/internal/PubsubIOTranslator.java

```diff
@@ -32,32 +34,25 @@
   /**
    * Implements PubsubIO Read translation for the Dataflow backend.
    */
-  public static class ReadTranslator<T> implements TransformTranslator<PubsubIO.Read.Bound<T>> {
+  public static class ReadTranslator<T> implements TransformTranslator<PubsubUnboundedSource<T>> {
     @Override
     @SuppressWarnings({"rawtypes", "unchecked"})
     public void translate(
-        PubsubIO.Read.Bound transform,
+        PubsubUnboundedSource<T> transform,
         TranslationContext context) {
       translateReadHelper(transform, context);
     }
 
     private <T> void translateReadHelper(
-        PubsubIO.Read.Bound<T> transform,
+        PubsubUnboundedSource<T> transform,
         TranslationContext context) {
-      if (!context.getPipelineOptions().isStreaming()) {
-        throw new IllegalArgumentException(
-            "PubsubIO.Read can only be used with the Dataflow streaming runner.");
-      }
+      Preconditions.checkState(context.getPipelineOptions().isStreaming(),
+          "PubsubIOTranslator.ReadTranslator is only for streaming pipelines.");
 
       context.addStep(transform, "ParallelRead");
       context.addInput(PropertyNames.FORMAT, "pubsub");
-      if (transform.getTopic() != null) {
-        context.addInput(PropertyNames.PUBSUB_TOPIC, transform.getTopic().asV1Beta1Path());
```

Comment:
Only the dataflow service has a notion of 'cleanup when the job is done', and so can delete random subscriptions created if PubsubIO.Read is only given a topic. So PubsubUnboundedSource requires a subscription and PubsubIO.Read requires it to be specified.

But then this translation approach won't work since the topic has not been captured by the PubsubUnboundedSource. Dang, missed that, thanks for noticing. Perhaps I can look at the parent PubsubIO.Read PTransform?